### ISSUE TITLE
 Search inside dictionary definition [Feature request]

### ISSUE THREAD

**Issue by Wacek** [koreader/koreader#3022]
Rationale for this: Many word definitions are extremly long, this is especially evident for example in English phrasal verbs. My paper printed English dicionarty has about 10 page long definition of all phrasals, usage examples and collocations for the word come (in two columns). I suppose the situation in my stardict dictionary is simililar. The problem is exacerbated by the fact, that the creators normally do not list all the phrasals alphabetically, but group them by meanig, but beacuse I obviously do not know the meaning beacuse I am using the dictionary this leads to the situation, where I have to read (or at least glance at) the entire text swiping it every time from beginning to end. I have already read the entire text for come probably a couple hundred  times in my life, but since I tend to forget the rarely used ones, I am forced to re-read this again and again. Similar situation is in other languages for example in German.

**Comment by KenMaltby** [comment]
How would a search be phrased to address an issue of a large amount of definition.  I guess if the definitions were grouped in categories (nautical terms, armorial terms, engineering terms, ect...) But I have found that it is more often that there are separate dictionaries produced, for such categories.  I understand that sdvc (the stardict application) is being improved upstream (in the project that it comes from).  As it is now, it is good at providing returns from multiple dictionaries.  That said, the dictionaries available (or practical) for use within an ereader are unlikely to resemble your "paper printed English dicionarty", especially as to size of the returned definitions. 

Not sure how such a search feature would work, or help youe described issue.  

Luck;
Ken

**Comment by Wacek** [comment]
You dont understand. Lets stick to the word come - then de definition lists all the phrasals, idioms collocations, regionalisms, slang terms under one entry COME. So that when you touch that word in the text you get as a recult:
Come:
a)xxxx 
- Come in xxxxx Come out xxxx Come inside xxxxx 
b) yyyy
-come off,yyy come round yyyy come up to yyyyy
Come aborad, come aroosss, come away, come back, come by 
idiom: 
all good things come to an end  etc, etc,
This is literally going on for page by page, again not alphabetically as a rule, but  rather by groupped by meaning. Reading through this every time yo stumble upon some unknown collocation to come is frustrating to the bone.

**Comment by KenMaltby** [comment]
How much is returned by a dictionary would depend on the dictionary itself, you might want to add a simpler dictionary and only use the more comprehensive dictionary when needed.  A large comprehensive dictionary (like the Oxford_English_Dictionary_2nd_Ed.) Will also slow down the lookup process.  

That said, have you tried?  You need to consider it with and without "fuzzy search" enabled.  I tried "might be" and got 20 results with fuzzy search on.  The third result was the proper one and was from a simpler dictionary.

**Comment by Wacek** [comment]
Fuzzy search is no point here, since there is no string for say "come apart" because everythin that has to do with come is under "come" and that's it. So to know what come apart means you need to read "come" definition thoroughly until you find g) ---come apart and that will take you some minutes to get there.
The dictionary is quite good, comprehensive, and usable, but also quite cumbersome simply because of its structure.
As for the dictionary change - yeah one can do that, and for English-English you have multiple options of course. But for some languages you can not find these dictionaries so easily, sometimes there is only one or two to choose from.
Anyway thanks for the response.

**Comment by KenMaltby** [comment]
OK, I enter come apart using the dictionary lookup (if I had in in an ebook I could longpress on come then slide over apart, before lifting my finger, and both words become what is searched for) and I get two dictionaries with a result:  The "WordNet" dictionary lists "come apart" and a valid definition.  The other dictionary; wikt-en-ALL-2015-07-20 also lists "come apart" with valid definitions.   You can query multiple words for an entry in the dictionaries.

**Comment by Wacek** [comment]
Sorry but I don't see your point here. I never said "there are no dictionaries out there that have these entries". Obviously there are, and that is not the issue . I simply say there is a group of dictionaries such as mine (and I have also German-Polish, and French-Polish from the same publishing house) which would benefit from having that feature (ie search within) because they are good and usable, but build in a cumbersome style. Thats it. If the feature is seen as not useful then I am ok wit that, just wanted to ask for it and see what comes out of it.

**Comment by KenMaltby** [comment]
I am only trying to understand the issue and how it would work as something differing from what is already there.  Given your last post I wonder if it may be that the existing multi-word lookup needs to delve deeper into the dictionaries.  Otherwise the change may need to be in the dictionaries.  
I would not be the one implementing such a feature, but the more detail understanding of the intent of the feature that can be established the more likely someone is to add it to their list.

**Comment by Wacek** [comment]
I am using it on GoldenDict desktop 1.5 for the same dictionaries (CTRL+F) and it just works. If you don't understand still I may send you the files so you could see yourself.

**Comment by Commodore64user** [comment]
with the OED (not another shameless plug), sometimes it is nigh impossible to visually find specific terms, something like this, i.e., search within the definition of an entry would be super helpful

**Comment by Commodore64user** [comment]
I’ve got a working prototype now. It only works with HTML dictionaries at the moment (the textbox is a pain in the arse, so I’ve left that for later). It does substring search and highlights all matches at once rather than one by one, and I’m basically trying to keep it as non-intrusive to DictQuick as possible since there’s no space left anywhere ;)

I’ll share some footage later. 👋

**Comment by Commodore64user** [comment]
One question i have is, I want to hijack either taps or swipe up/down to cycle through pages with matches while keeping the other moving through all pages (when in this search mode only of course). Which makes more sense to keep as advance one page at a time and which for next page with matches?

On `hasKeyboard` i am using the <kbd>.</kbd> and <kbd>,</kbd> keys, not sure what to do for K4 though.

**Comment by Commodore64user** [comment]
the button is the only visual addition, it takes the whole row to itself or shares it with presets if there are any.

there are no options or settings to tweak, is substring so it meets most of everyone's demands whilst keeping it nice and cheap. As you scroll, all the matches found in that page will be highlighted. As mentioned in the previous message as well, ideally either tap or swipes should move on a per matched page only basis. Any comments from @poire-z @offset-torque @Frenzie @mergen3107 @jonnyl2 @hius07 or any Linguist-to-be?

<p align="center">
<img src="https://github.com/user-attachments/assets/75f6931c-ca35-4fe2-a66a-7b48175c2314" width=40%> 
<img src="https://github.com/user-attachments/assets/c4165245-5119-4558-a0d2-a4a6d9fc4613" width=40%> 
</p>

**Comment by Frenzie** [comment]
The button is a bit ugly, but I suppose it can't be helped.

>  substring so it meets most of everyone's demands whilst keeping it nice and cheap

They're usually case insensitive, but that kind of thing isn't terribly important initially.

**Comment by jonnyl2** [comment]
Why not use this kind of ButtonDialog to jump between matches that users are familiar with from Fulltext Search? (Probably without the middle button.)

![Image](https://github.com/user-attachments/assets/17e6aaf5-f5fc-4fd6-869a-d96a434ac299)

https://github.com/koreader/koreader/blob/17281b2f1fe700d7a8b64295d861e103e2ac45ee/frontend/apps/reader/modules/readersearch.lua#L486

**Comment by poire-z** [comment]
> As you scroll, all the matches found in that page will be highlighted. As mentioned in the previous message as well, ideally either tap or swipes should move on a per matched page only basis

I use tap to scroll up/down. I guess that after I have call this feature to find in defintion, I would somehow expect tap to lead me to the matches - and my less used swipe up/down could allow to read the page up/down where there may be no match.
But that's just me and people using swipes might want the reverse, or think different :)

> Why not use this kind of ButtonDialog to jump between matches 

Because it's more complicated ? :)

> On `hasKeyboard` i am using the . and , keys

Damned Britocentric !
<img width="498" height="170" alt="Image" src="https://github.com/user-attachments/assets/987f36a7-383e-467d-a7e3-94d787d3baea" />

**Comment by jonnyl2** [comment]
>[Commodore64user] ideally either tap or swipes should move on a per matched page only basis

How does one exit out of the search result highlights and restore the regular definition view?

**Comment by Commodore64user** [comment]
<p align="center">
<img src="https://github.com/user-attachments/assets/5f7cbd11-39e6-454a-b748-dce6d61d1487" width=40%> 
</p> 

with presets it looks a bit less ugly, but I suppose having three rows is the offender here, alas what could be done? (ps. I never translate or use wiki from there... just saying)

> They're usually case insensitive

it is ;)

> Why not use this kind of ButtonDialog to jump between matches that users are familiar with from Fulltext Search? (Probably without the middle button.)

because it makes the whole thing far more complex, this is meant for getting to specific matches quick, not for an in-depth investigation of etymology and word use in 15th century Britain ;)

> How does one exit out of the search result highlights and restore the regular definition view?

move to the next result (i.e., 2/12 in this case) or start a new "find in definition" or do a long press to start a new dictionary search, or leave the widget altogether. I think that is it.

lol
<p align="center">
<img src="https://github.com/user-attachments/assets/2bd39e13-23ec-41cf-8502-98c8b9de3c26" width=60%> 
</p>

**Comment by jonnyl2** [comment]
Ok. I admit I have wished to have this feature available before, so it's definitely nice to have.

Another question: while results are being displayed, is jumping between pages with results the only navigation available? Or is regular page-by-page navigation still possible?

**Comment by Commodore64user** [comment]
Somebody get this guy some reading glasses 🤓 

Re-read the nine years worth of messages ;)

**Comment by jonnyl2** [comment]
My bad. My vote is: hijack swipe north/south to move between in-definition results. Taps to page within definition as usual. But what happens at the end of the article? It probably should not jump to the next article while in search mode.

**Comment by Commodore64user** [comment]
Another question, if one is in said search mode, would you prefer when tapping the `x` in the top corner or the `Close` button to either: exit said mode only and for the button to go back to its normal behaviour OR to directly quit the widget (business as usual)?

**Comment by Frenzie** [comment]
I'd say just quit.

**Comment by poire-z** [comment]
Also just quit (although if I had done this feature, I would probably be lazy and use that to cancel it, and I would argue it's a good idea :)).
May be you could morph the Edit IconWidget (that pencil right next to 1/12) into another icon, and press on it would quit that mode?

**Comment by Commodore64user** [comment]
> May be you could morph the Edit IconWidget (that pencil right next to 1/12) into another icon, and press on it would quit that mode?

Is not one of those, "is `x`'s all the way down" is it? ("is turtles all the way down" for anyone who doesn't get the ref)

**Comment by poire-z** [comment]
oh, may is of these, isn' ?

**Comment by Commodore64user** [comment]
You know it fam, innit

**Comment by offset-torque** [comment]
In that window, there is a scroll bar. Users seeing this will probably expect the "swipe up/down" to scroll the page (experience/muscle memory from phones, tablets...). So this leaves "tap" for the other option (cycle through pages with matches).

Also tap is a bit quicker and less effort so if I am jumping between lots of results, I would prefer 10 taps to 10 swipes.

No comment on "quit", as I am not familiar with the current dictionary workflow that users are accustomed to.

**Comment by Commodore64user** [comment]
> In that window, there is a scroll bar. Users seeing this will probably expect the "swipe up/down" to scroll the page (experience/muscle memory from phones, tablets...). So this leaves "tap" for the other option (cycle through pages with matches).

incidentally this is one hundred percent the reason why I already used taps for pages with matches and definitely not because `onTap` was slightly easier to hijack ;)

**Comment by Commodore64user** [comment]
> if I had done this feature, I would probably be lazy and use that to cancel it, and I would argue

btw @poire-z if you wanted to do the `TextBoxWidget` version... I wouldn't be against it ;)

**Comment by poire-z** [comment]
am not one of those

**Comment by Commodore64user** [comment]
well in that case, TextBoxWidget is going to have to be delayed to 2026.04.1 (probably more like 2026.04.3)

**Comment by Commodore64user** [comment]
if you have a kobo libra or kindle oasis, or any device with a single pair of page turns buttons, would you prefer 'em to loop through pages with matches OR business as usual? I am inclined to say, loop through matches...

**Comment by jonnyl2** [comment]
Intuitively I would say loop through pages with matches, but it really depends what tap-on-screen does. Ideally it would be the other thing, so both navigation options could be controlled without needing to swipe.

**Comment by Commodore64user** [comment]
what I am asking essentially is, on devices with two sets, should I use Left or Right set for pages with matches. If I use the right, that will affect all those devices with a single pair as they are usually labeled as the Right set, even though there is only one set. So forget about what onTap does, even though we already know what it does https://github.com/koreader/koreader/issues/3022#issuecomment-3910647340 ;)

**Comment by jonnyl2** [comment]
Then I'd say use Right buttons for regular paging (business as usual). I can't say what to do with the Left set though. I don't know if users expect those to have the same function (and just use those when they hold the device in the other hand).

**Comment by mergen3107** [comment]
(Admittedly, I don't have a buttoned device anymore)
You guys might know my stance at this point. 
1) Whatever new changes are, it should default to old behavior.
2) New changes should be all configurable.
3) New behavior should be announced via popups or long press info messages or the User Guide, or all of those :D 

Let me explain how I see use cases so you could understand why I say having it configurable is important:

Common starting point:
A user searches a common word
Popup shows a very long dictionary article. 
User searches for a specific word in that article and finds N matches (by the way, how about showing how many matches were found after search?)

Now, Case A)
User wants to use buttons to scroll through the article quickly to see in which context in the article this word appears. I say quickly here because intuitively while reading a book buttons suggest "page turns" and not "jump between search results". So old behavior memory continues here.
Then user can use taps/swipes (again, configurable) to jump between found words. 

Case B)
Popup appears warning that buttons are now used to jump between search results. Taps/swipes are configured on their own. 

What I would say though is that at least one type of input (buttons, swipes, taps) should stay default no matter what. Otherwise expect foot gun situations from not fully aware users :D 

P.S. In my mind, perfectly there might be two solutions, but both of those I guess are blocked by limitations of the dictionary widget:
1) Like in some editors (Matlab) when you search and find stuff, their relative positions are highlighted **in the** scroll bar. Visual users know where appropriately to expect them.
2) When a search completes, two buttons appear at the top and bottom of the scroll bar (or next to it): Prev/Next found word. To jump between them, users just need to tap those new buttons. Kinda what Notepad++ (or other text editors) to "Search to find next" and the same with "Backward direction" enabled. 

OR the same way skim widget jumps between the "stars" (bookmarks/notes/highlights). By the way, can you borrow any logic and code from there? Might save you some time. Or maybe replicate a mini skim widget just for the dictionary widget?

**Comment by Commodore64user** [comment]
### The Case Against Configurable:

I remember an old time legend (might have been me ;)) once said: "sometimes is not just about what you do, but what you don't do". Same concept applies here, is not about what is being added but what is being left out. You have already correctly pointed out a few issues, there are constraints, unlike normal search, this lives inside a widget so thinking of it like the search feature for a document is just plainly incorrect.

I am bound by the `DictQuickLookUp` widget, both in terms of what it can do and its ease of operability. So saying "simply pop up another widget" is not only the wrong approach but also fails to understand the real needs of why this is being added in the first place. This is meant to find specific terms within long entries much easily, that is it. Dictionary entries don't tend to be much longer than perhaps 10 pages, sure the OED has lots of exceptions to that rule but for the most part, this is sufficient for purpose.

But my browser can pop up the button for next/previous matches. Sure it can, but it also has plenty of space, which `DictQuickLookUp` does not, besides I have to make sure that all other functionality remains accessible to NT devices as well. Even if I ignore this, what's the point?, why would I add some buttons that cycle through every match? look at the screenshot in https://github.com/koreader/koreader/issues/3022#issuecomment-3905266110 an tell me, why have some extra buttons that move through the matches one at a time? for what valid reason should I complicate my life like that? there is no real benefit to that here.

user should decide if taps, or swipe cycle through matches. Hard disagree there (more complexity with little payoff), first of all, nothing is changing both Taps and Swipes will continue to work in much the same they do, it is only after one deliberately searches that Taps change their behaviour to loop through the pages with matches, furthermore a nice new section shall appear on the user guide ;). 

so in short, we can't have widget all the way down. I have stripped down the search to its basic form where is both still useful and non-invasive to its host.

**Comment by Frenzie** [comment]
1, 2, and 3 are all greatly situational. I don't think it makes much sense to make something like search highlights super configurable. That just makes the code impossible to maintain.

Also don't you get the old behavior by default regardless? Don't search for anything = you don't get any search highlights with associated search highlights behavior.

**Comment by mergen3107** [comment]
OK :D 
Then super simple solution:
1) Widget behaves normally when first shown on lookup
2) If you search (one extra button), then all controls only move between search results. 
3) If you want to casually scroll again, exit out of the search "mode".

**Comment by mergen3107** [comment]
I mean, this feature (search within lookups) is super niche in itself anyway (only linguistic can appreciate it fully, I guess). So I don't get it when you guys jump on my discussion of maintaining this super niche regime :D

**Comment by poire-z** [comment]
> This is meant to find specific terms within long entries much easily, that is it. Dictionary entries don't tend to be much longer than perhaps 10 pages, sure the OED has lots of exceptions to that rule but for the most part, this is sufficient for purpose.

Then, why not just drop the idea of changing gestures to navigate to page with matches ? If they are less than 10 pages, the user can just tap tap tap, the highlighted text will just catch his eyes (and while browsing down, he will glance at the context; if jumped directly, he may probably have to scroll up to see that context).

**Comment by Commodore64user** [comment]
> 10 pages

Unless you are reading the OED, where you might get quite a lot longer entries ;). I believe the longer entry is "set" (pos: verb) which has north of 200 pages 😬. Tap tap tap won't do there ;)

**Comment by poire-z** [comment]
But what will you search for in such a "set" definition ?

**Comment by jonnyl2** [comment]
It is conceivable that you want to search for some idiom containing "set". But then there should only be one result, maybe two or three (but right after one another).  As long as the search jumps to the first search result, there shouldn't be much paging between results needed anymore.

**Comment by Frenzie** [comment]
> OK :D Then super simple solution:
> 
>     1. Widget behaves normally when first shown on lookup
> 
>     2. If you search (one extra button), then all controls only move between search results.
> 
>     3. If you want to casually scroll again, exit out of the search "mode".

I thought that was more or less the proposal, yes. The question of what to do about Kindles with two sets of page turn buttons is orthogonal.

**Comment by jonnyl2** [comment]
But what if a search result is on the last line of a screen page? Then I'd have to exit out of the search mode just to read the remaining part of the definition.

**Comment by Frenzie** [comment]
I wouldn't do 2, no. I'd simply make it different behavior between swipe left/right vs up/down. My point is merely that the left/right swipe (or whatever) would not be activated unless you search, so there is nothing backwards incompatible about the behavior.

**Comment by Commodore64user** [comment]
> But what will you search for in such a "set" definition ?

<p align="center">
<img src="https://github.com/user-attachments/assets/727ea27c-4669-40e0-b7f4-dd95f22dd57b" width=40%> 
<img src="https://github.com/user-attachments/assets/811fc914-097f-44f7-b41b-a1e29faf1b6c" width=40%> 
</p> 

> But then there should only be one result, maybe two or three (but right after one another). As long as the search jumps to the first search result, there shouldn't be much paging between results needed anymore.

incorrect.
on the left is the first ever match (well the second actually, first is from a quotation), on the right, what one might actually want to read.. those are about a hundred pages apart.

> I wouldn't do 2,

and we are clear that ain't happening. only taps change, swipes remain the same.

**Comment by offset-torque** [comment]
> furthermore a nice new section shall appear on the user guide ;).

Added to my task list. Are there other interactions that needs to be documented similar to this in dictionary/wikipedia widgets?

This is currently the only one in the guide: **"Long press on close icon to close all windows (dictionary/wikipedia)"**

**Comment by Commodore64user** [comment]
> Added to my task list. Are there other interactions that needs to be documented similar to this in dictionary/wikipedia widgets?

unrelated to this, but the other day I was thinking the whole "presets" business could be explained in the guide. (status bar presets and dictionary presets)

3 second long press on the dictionary widget will search in the opposite to where you are dictionary-wikipedia, on NT is called by pressing modifier+press during text selection.

**Comment by Commodore64user** [comment]
@poire-z or @Frenzie  could you please grab an html dictionary you have, and change its .`ifo` from `h` to `x`. Then open the dictquicklookup and invoke text_selection with <kbd>shift</kbd> + <kbd>up</kbd>, try and select anything, I get this... something is messing with the coordinates... seems to select just fine when on wikipedia though 

<img src="https://github.com/user-attachments/assets/9998d883-6de7-408d-bb9f-fa0ccc1d7a45" width=40%>

**Comment by poire-z** [comment]
I got the issue on an existing =x or =m:
When using the mouse pointer to select "Une maison":

<img width="532" height="404" alt="Image" src="https://github.com/user-attachments/assets/c82d7764-be07-4257-b75c-543fcea22d4b" />

When using Shift+Up to select the same thing:

<img width="529" height="406" alt="Image" src="https://github.com/user-attachments/assets/8bdb14b5-9464-4f50-8618-8da365acfb11" />

Same issue for me on a Wikipedia full article:

<img width="588" height="344" alt="Image" src="https://github.com/user-attachments/assets/a0d4a112-c65c-47ad-ad72-bb5a2e094149" />

**Comment by Commodore64user** [comment]
I tried the wiki non-full and that one seemed fine, any ideas who's acting up?

**Comment by poire-z** [comment]
Nope (busy with other things). (And I won't take from you the joy of debugging and figuring it out yourself :))

**Comment by Commodore64user** [comment]
What other things? Fix thy textboxwidget ;)

**Comment by Commodore64user** [comment]
kindles on `2025.10-57-g3644299a6_2025-12-12` and `2025.10-117-g642ec66d6_2026-02-06` don't seem to have that problem on both full wiki and non full wiki... html also doesn't have that problem.

**Comment by Commodore64user** [comment]
what do all your screenshots and mine have in common? and my wikipedia entry for dog doesn't... utf8.. character indices vs byte indices strike again

**Comment by poire-z** [comment]
> kindles on `2025.10-57-g3644299a6_2025-12-12` and `2025.10-117-g642ec66d6_2026-02-06` don't seem to have that problem on both full wiki and non full wiki... html also doesn't have that problem.

Was that really true ? The code you adjusted was 1 year old.

**Comment by Frenzie** [comment]
There was a minor MuPDF update.

https://github.com/koreader/koreader-base/commit/f45e77be4639c7f077cd81f16a2874ba468a85b1

**Comment by Commodore64user** [comment]
> > kindles on `2025.10-57-g3644299a6_2025-12-12` and `2025.10-117-g642ec66d6_2026-02-06` don't seem to have that problem on both full wiki and non full wiki... html also doesn't have that problem.
> 
> Was that really true ? The code you adjusted was 1 year old.

By this point you probably already have the answer to this... but in my tests, the wikis i opened only had ascii characters, therefore no shift. The bug has been there all along

**Comment by TnS-hun** [comment]
From @Commodore64user:
> start a new "find in definition"

Then only way to cancel it this way is to enter a search term that cannot be found.

> do a long press to start a new dictionary search, or leave the widget altogether

Neither of them is too convenient.

> move to the next result (i.e., 2/12 in this case)

So basically this is the only easy way to cancel a search, but if you are using only a single dictionary or there are no more results in other dicitionaries then you are out of luck.

And what if you want to search for the same thing within the definitions in another dictionary after clicking to the >> button? You have to enter search term again. If there would be a better way to cancel the search then the search could automatically continue after pressing the <<  >> buttons.

**Comment by Commodore64user** [comment]
for anyone interested, since base was bumped recently by the one and only @benoit-pierre, you can test on your device now, should you want to... on challenging entries (set) looking for `a` seems to happen in about 5-10 secs (pw5), which seems reasonable for such a long entry. Shorter entries should yield a much faster response.

incidentally, for TextBox (wiki) we cache the original lowercasing which should speed up significantly subsequent searches on the same entry, but that isn't ready yet.

**Comment by Commodore64user** [comment]
@poire-z save the date

<p align="center">
<img src="https://github.com/user-attachments/assets/78900f5e-5272-4301-ad33-ee55fab86482" width=40%> 
</p>
